docker/beam-environment/entrypoint.sh

```diff
@@ -0,0 +1,362 @@
+  vval1="$v"
+  vval="${!vval1}"
+  echo "$v -> '$vval'"
+done
+
+
+##
+## fix max ram if not set
+##
+if [[ -z "$MAX_RAM" ]]; then MAX_RAM="8"; echo "MAX_RAM was not set, using default value: '$MAX_RAM'"; fi
+
+
+##
+## print optional variables that might be set only outside of image
+##
+optional_variables_from_outside=( PULL_CODE PULL_DATA SEND_NOTIFICATION )
+echo "Following OPTIONAL variables might be set only outside of the image (variable name -> 'current value'):"
+for v in "${optional_variables_from_outside[@]}" ; do
+  vval1="$v"
+  vval="${!vval1}"
+  if [[ -z "$vval" ]]; then
+    eval "$v=true"
+    vval1="$v"
+    vval="${!vval1}"
+    echo "$v -> '$vval'  (using default value)"
+  else
+    echo "$v -> '$vval'"
+  fi
+done
+
+
+function send_slack_notification() {
+  if [[ -z "$SLACK_HOOK_WITH_TOKEN" ]]; then
+    echo "WARNING: Can't send notifications to slack, SLACK_HOOK_WITH_TOKEN is not set!"
+  else
+    json_data="{\"text\":\"$1\"}"
+    printf "\nSending the following json to slack:"
+    echo "$json_data"
+    echo " "
+    curl -X POST -H 'Content-type:application/json' --data "$json_data" "$SLACK_HOOK_WITH_TOKEN"
+    echo " "
+  fi
+}
+
+function send_json_to_spreadsheet() {
+  if [[ -z "$SIMULATIONS_SPREADSHEET_UPDATE_URL" ]]; then
+    echo "WARNING: Can't send updates to the spreadsheet, SIMULATIONS_SPREADSHEET_UPDATE_URL is not set!"
+  else
+    json_data="{\"command\":\"add\",\"type\":\"beam\",\"run\":{ $1 } }"
+    printf "\nSending the following json to the spreadsheet:"
+    echo "$json_data"
+    echo " "
+    curl -X POST -H 'Content-type:application/json' --data "$json_data" "$SIMULATIONS_SPREADSHEET_UPDATE_URL"
+    echo " "
+  fi
+}
+
+
+##
+## logging CPU | RAM usage during simulation
+##
+CPU_RAM_LOG="/app/sources/cpu_ram_usage.csv"
+/app/write_cpu_ram_usage.sh > "$CPU_RAM_LOG" &
+
+
+##
+## location to project folder
+##
+PATH_TO_PROJECT_PARENT="/app/sources"
+cd "$PATH_TO_PROJECT_PARENT" || echo "ERROR: The path '$PATH_TO_PROJECT_PARENT' is not available"
+
+
+##
+## either pull the code or use the code from mounted folder
+##
+if [ "$PULL_CODE" = true ]; then
+  BEAM_NAME="beam"
+  if [ -z "$BEAM_CODE_URL" ]; then BEAM_CODE_URL="https://github.com/LBNL-UCB-STI/beam.git"; fi
+  echo "Pulling the code from github (PULL_CODE set to '$PULL_CODE'), cloning '$BEAM_CODE_URL' into $(pwd)/$BEAM_NAME"
+  git clone --single-branch --branch "$BEAM_BRANCH_NAME" "$BEAM_CODE_URL" "$BEAM_NAME"
+  cd "$BEAM_NAME" || echo "ERROR: The dir '$BEAM_NAME' is not available"
+
+  if [[ $BEAM_COMMIT_SHA ]]; then
+    echo "Resetting the branch '$BEAM_BRANCH_NAME' to commit '$BEAM_COMMIT_SHA'"
+    git reset --hard "$BEAM_COMMIT_SHA";
+  else
+    BEAM_COMMIT_SHA=$(git log -1 --pretty=format:%H)
+    echo "Using the latest commit in the branch '$BEAM_BRANCH_NAME' - '$BEAM_COMMIT_SHA'"
+  fi
+else
+  echo "Pulling the code is disabled (PULL_CODE set to '$PULL_CODE'), using code mounted to '$PATH_TO_PROJECT_PARENT'."
```

Comment:
I agree there might be some confusion in paths when the code is pulled and when not - there will be an additional level of folders when git creates a new directory to pull in there.
I was thinking about that and haven't found anything simple and more intuitive for user than what I already did because:
1. removing folders does not sound like a good idea - there might be something inside,
2. existing beam code might be inside of a folder with any name, not exactly `beam`
3. one of requirements was to be able to mount data and code separately and they both might be in the same parent folder.

There will be also some kind of confusion with path to config when data folder was mounted because parent folder name is lost when mounted, but again - I see no clear and intuitive solution for that.

Reply:
Good thing you pointed at this, it appears there is a better solution - to git clone into current folder, which is simplifying paths and reduces confusion.

Reply:
This still doesn't check whether there is something existing in the source and data directory as far as I can tell. So, if you are relying on it being there then it might not..especially since there is no dockerfile for a builtin scenario

Reply:
I expect the mounted folders will either contain code\data or any of `pull` flags to be true, if neither - then the running of the image will fail